Add more details to labels model in 5.0 guide (#9091)
GoogleCloudPlatform/magic-modules: mmv1/third_party/terraform/website/docs/guides/version_5_upgrade.html.markdown

* The output-only `terraform_labels` will merge the labels defined by the users on the resource through Terraform and the default labels configured on the provider. If the same label exists on both the resource labels and provider default labels, the label on the resource will override the provider label.
* The output-only `effective_labels` will list all of labels present on the resource in GCP, including the labels configured through Terraform, the system, and other clients.

**Note:** `ignore_changes` can be applied to `labels` field to ignore the changes of the user defined labels. It is not recommended to apply `ignore_changes` to `terraform_labels` or `effective_labels`, as it may unintuitively affect the final API call.

The following changes will be observed after upgrading to `5.0.0`.

> * Running `terraform import` on Google provider `5.0.0` or later, these three fields will show in the state file with an empty `labels` and `terraform_labels` value. `effective_labels` will have all of labels present on the resource in GCP. You can update the resource to bring labels back under management.

Reviewer: ```suggestion
* Running `terraform import` on Google provider `5.0.0` or later, these three fields will show in the state file with an empty `labels` and `terraform_labels` value. `effective_labels` will have all of labels present on the resource in GCP. You can update the resource to bring labels defined in your configuration under management by Terraform.
```

> * Running `terraform refresh` on Google provider `5.0.0` or later, these three fields will show in the state file. `labels` field will have your current labels, `terraform_labels` will be empty, and `effective_labels` will have all of labels present on the resource in GCP.

Reviewer: ```suggestion
* Running `terraform refresh` on Google provider `5.0.0` or later with an existing pre-`5.0.0` resource before an `apply`, these three fields will show in the state file. `labels` field will have your current labels, `terraform_labels` will be empty, and `effective_labels` will have all of labels present on the resource in GCP.
```

> * Running `terraform plan` or `terraform apply` on Google provider `5.0.0` or later, the plan will show an updated adding your current labels to `terraform_labels`. After running `terraform apply`, these three fields will show in the state file. `labels` will have your current labels, `terraform_labels` will have the combination of `labels` and your provider-default labels, and `effective_labels` will have all of labels present on the resource in GCP.

Reviewer: ```suggestion
* Running `terraform plan` or `terraform apply` on Google provider `5.0.0` or later with an existing pre-`5.0.0` resource before an `apply`, the plan will show an updated adding your current labels to `terraform_labels`. After running `terraform apply`, these three fields will show in the state file. `labels` will have your current labels, `terraform_labels` will have the combination of `labels` and your provider-default labels, and `effective_labels` will have all of labels present on the resource in GCP.
```

> * In the resource `google_cloud_run_domain_mapping`, the system labels `cloud.googleapis.com/location` and `run.googleapis.com/overrideAt` are removed from `labels` inside `metadata` field in the state file. If any of these label keys are in the configuration, after upgrading to `5.0.0`, the plan shows that these keys will be added. You can safely accept this change.

Reviewer: ```suggestion
* In the resource `google_cloud_run_domain_mapping`, the system labels `cloud.googleapis.com/location` and `run.googleapis.com/overrideAt` will be removed from `labels` inside `metadata` field in the state file as part of a one-time resource schema upgrade. If any of these label keys are in the configuration after upgrading to `5.0.0`, the plan will show these keys will be added. You can safely accept this change, and Terraform will begin to manage them again.
```

* In the resource `google_cloud_run_service`, the system annotations `serving.knative.dev/creator`, `serving.knative.dev/lastModifier`, `run.googleapis.com/operation-id`, `run.googleapis.com/ingress-status`, `run.googleapis.com/ingress` with the value `all` are removed from `annotations` inside `metadata` field in the state file. If any of these annotation keys are in the configuration, after upgrading to `5.0.0`, the plan shows that these keys will be added. You can safely accept this change.

#### Data source annotations

For most resource-based datasources, both `annotations` and `effective_annotations` will now be present. Both fields will have all of annotations present on the resource in GCP including the annotations configured through Terraform, the system, and other clients, equivalent to `effective_annotations` on the resource.

#### Example
##### Config
```hcl
provider "google" {
  default_labels = {
    default_key = "default_value"
  }
}

resource "google_dataproc_cluster" "with_labels" {
  name   = "tf-test-dproc-test-1"
  region = "us-central1"

  labels = {
    key1 = "value1"
    key2 = "value2"
  }
}
```

> After the configration is applied, Terraform is managing `key1` and `key2` in the `labels` field. `terraform_labels` field has label `default_key`, `key1` and `key2`. `effective_labels` has label `default_key`, `key1`, `key2` and system labels.

Reviewer: ```suggestion
After the configuration is applied, Terraform is managing `key1` and `key2` in the `labels` field. `terraform_labels` field has label `default_key`, `key1` and `key2`. `effective_labels` has label `default_key`, `key1`, `key2` and system labels.
```

##### Config
```hcl
provider "google" {
  default_labels = {
    default_key = "default_value"
  }
}

resource "google_dataproc_cluster" "with_labels" {
  name   = "tf-test-dproc-test-1"
  region = "us-central1"

  labels = {
    key1 = "value1"
  }
}
```

> After the configration is applied, Terraform stops managing `key2` and is managing `key1` in the `labels` field. `terraform_labels` field has label `default_key` and `key1`. `effective_labels` has label `default_key`, `key1` and system labels.

Reviewer: ```suggestion
After the configuration is applied, Terraform stops managing `key2` and is managing `key1` in the `labels` field. `terraform_labels` field has label `default_key` and `key1`. `effective_labels` has label `default_key`, `key1` and system labels.
```

* The output-only `terraform_labels` will merge the labels defined by the users on the resource through Terraform and the default labels configured on the provider. If the same label exists on both the resource labels and provider default labels, the label on the resource will override the provider label.
* The output-only `effective_labels` will list all of labels present on the resource in GCP, including the labels configured through Terraform, other clients and services.

**Note:** `ignore_changes` can be applied to `labels` field to ignore the changes of the user defined labels. It is not recommended to apply `ignore_changes` to `terraform_labels` or `effective_labels`, as it will make the new labels model not work as expected.

The following changes will be observed after upgrading to `5.0.0`.
* After upgrading to `5.0.0`, and then running `terraform import`, these three fields should show in the state file of the resources with empty `labels` and empty `terraform_labels`. `effective_labels` will have all of labels present on the resource in GCP.

> * After upgrading to `5.0.0`, and then running `terraform refresh`, these three fields should show in the state file of the resources with a self-applying `labels` field and empty `terraform_labels`. `effective_labels` will have all of labels present on the resource in GCP.

Reviewer: "with a self-applying `labels` field" what does this mean?

I think we expect to see the current set of labels in state preserved as the value for `labels`, right?

Author: that is right. Changed.

> * After upgrading to `5.0.0`, and then running `terraform plan` or `terraform apply`, the plan shows that a new block `terraform_labels` will be added to the state. After running `terraform apply`, these three fields should show in the state file of the resources with a self-applying `labels` field. In the state file, `terraform_labels` will have the merged labels and `effective_labels` will have all of labels present on the resource in GCP.

Reviewer: ```suggestion
* After upgrading to `5.0.0`, and then running `terraform plan` or `terraform apply`, the plan will show an updated adding your current labels to `terraform_labels`. After running `terraform apply`, these three fields should show in the state file of the resources with a self-applying `labels` field. In the state file, `terraform_labels` will have the combination of `labels` and your provider-default labels and `effective_labels` will have all of labels present on the resource in GCP.
```

> * In the resource `google_cloud_run_domain_mapping`, the system labels `cloud.googleapis.com/location` and `run.googleapis.com/overrideAt` are removed from `labels` inside `metadata` field in the state file. If any of these label keys are in the configration, after upgrading to `5.0.0`, the plan shows that these keys will be added.

Reviewer: ```suggestion
* In the resource `google_cloud_run_domain_mapping`, the system labels `cloud.googleapis.com/location` and `run.googleapis.com/overrideAt` are removed from `labels` inside `metadata` field in the state file. If any of these label keys are in the configuration, after upgrading to `5.0.0`, the plan shows that these keys will be added. You can safely accept this change.
```

Reviewer: (same for others below)

* In the resource `google_dataflow_job`, the system labels with the prefix `goog-dataflow-provided` are removed from the `labels` field in the state file. If any of these system label keys are in the configration, after upgrading to `5.0.0`, the plan shows that the keys will be added.
* In the resource `google_dataproc_cluster`, the system labels with the prefix `goog-dataproc` are removed from the `labels` field in the state file. If any of these system label keys are in the configration, after upgrading to `5.0.0`, the plan shows that these keys will be added.
* In the resource `google_notebooks_instance`, the system labels with the prefix `goog-caip-notebook` are removed from the `labels` field in the state file. If any of these system label keys are in the configration, after upgrading to `5.0.0`, the plan shows that these keys will be added.
* In the resource `google_storage_bucket`, the system labels with the prefix `goog-dataplex` are removed from the `labels` field in the state file. If any of these system label keys are in the configration, after upgrading to `5.0.0`, the plan shows that these keys will be added.

#### Data source labels

> Since the resource has used the new labels model, the corresponding data source now has three label-related fields as well, `labels`, `effective_labels` and `terraform_labels`. All of these three fields will have all of labels present on the resource in GCP, including the labels configured through Terraform, other clients and services.

Reviewer: ```suggestion
For most resource-based datasources, all three of `labels`, `effective_labels` and `terraform_labels` will now be present. All of these three fields will have all of labels present on the resource in GCP including the labels configured through Terraform, the system, and other clients, equivalent to `effective_labels` on the resource.
```
